src/builder.test.ts

```diff
@@ -0,0 +1,398 @@
+      q("*")
+        .filter("_type == 'pokemon'")
+        .slice(0, 3)
+        .grab({ name: q.string() })
+        .grab({
+          foo: ["name", q.string()],
+          // let's also query something that doesn't exist – groq will null it out
+          bar: q.string().nullable(),
+        })
+    );
+
+    expect(query).toBe(`*[_type == 'pokemon'][0..3]{name}{"foo": name, bar}`);
+    invariant(data);
+    expect(data[0].foo).toBe("Bulbasaur");
+    expect(data[0].bar).toBeNull();
+  });
+});
+
+describe("UnknownResult.filter/ArrayResult.filter", () => {
+  it("applies simple filter appropriately to UnknownResult, returning unknown array", async () => {
+    const { query, schema, data } = await runPokemonQuery(
+      q("*").filter("_type == 'pokemon'")
+    );
+
+    expect(query).toBe(`*[_type == 'pokemon']`);
+    expect(schema instanceof z.ZodArray).toBeTruthy();
+    expect(schema.element instanceof z.ZodUnknown).toBeTruthy();
+    expect(Array.isArray(data) && "name" in data[0]).toBeTruthy();
+  });
+
+  it("can stack filters, with no projection, schema is still unknown", async () => {
+    const { query, schema, data } = await runPokemonQuery(
+      q("*").filter("_type == 'pokemon'").filter("name match 'char*'")
+    );
+
+    expect(query).toBe(`*[_type == 'pokemon'][name match 'char*']`);
+    expect(schema instanceof z.ZodArray).toBeTruthy();
+    expect(schema.element instanceof z.ZodUnknown).toBeTruthy();
+    expect(data?.length).toBe(3);
+  });
+
+  it("can filter, project, and filter, and schema of projection is preserved", async () => {
+    const { query, schema, data } = await runPokemonQuery(
+      q("*")
+        .filter("_type == 'pokemon'")
+        .grab({ name: z.string() })
+        .filter("name match 'char*'")
+    );
+
+    expect(query).toBe(`*[_type == 'pokemon']{name}[name match 'char*']`);
+    expect(schema instanceof z.ZodArray).toBeTruthy();
+    expect(schema.element instanceof z.ZodObject).toBeTruthy();
+    invariant(data);
+    expect(data[0].name).toEqual("Charmander");
+  });
+
+  it("turns unknown schema into unknown array schema", async () => {
+    const { schema: schema1 } = await runPokemonQuery(q(""));
+    expect(schema1 instanceof z.ZodUnknown);
+
+    const { schema: schema2 } = await runPokemonQuery(
+      q("").filter("_type == 'animal'")
+    );
+    expect(
+      schema2 instanceof z.ZodArray && schema2.element instanceof z.ZodUnknown
+    );
+  });
+});
+
+describe("ArrayResult.order", () => {
+  it("applies order, preserves unknown schema", async () => {
+    const { query, schema, data } = await runPokemonQuery(
+      q("*").filter("_type == 'pokemon'").order("name asc")
+    );
+
+    expect(query).toBe(`*[_type == 'pokemon']|order(name asc)`);
+    expect(schema instanceof z.ZodArray).toBeTruthy();
+    expect(schema.element instanceof z.ZodUnknown).toBeTruthy();
+    // @ts-expect-error data is unknown type since no grab present
+    expect(data?.[0]?.name).toBe("Abra");
```

Comment:
hmm, I'm not sure I follow. Right now `data` has a type of `unknown[]` since no `grab` call was made. My line of thinking here is: TS doesn't have a way to know that `data` is an array of objects that have a `name: string` field, so we'd expect TS to balk. I'm not sure we'd want to force a type on `data`, because the goal of this test case is to show that there is no (useful) type without a `grab` or `grabOne` call.

Reply:
Ahhh gotcha! I see what you're saying. I hadn't thought much about that, because bypassing the Zod validation (by skipping the `.grab` calls) eliminates a lot of the utility of the library. But I'd say your best bet is just casting, exactly as you have shown above.